### PR TITLE
Fix issues with `dynamic` in pattern-matching

### LINKED ISSUE #17089: Declaration pattern should not support `dynamic`
Since no runtime object ever has type `dynamic`, it should not be supported as the type in a declaration pattern. The compiler does, today, permit that.

Related to that, when the input is of type `dynamic`, the compiler does not currently permit any type in a declaration pattern (other than `var` or `dynamic`). The compiler should treat such an input as being of type `object`.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #17089 

@AlekseyTs @agocke Please review

@dotnet/roslyn-compat-council This is technically a breaking change, as we now forbid `dynamic` as a pattern type (no object ever has `dynamic` as its runtime type).

### CODE REVIEW

**Comment by gafter** [comment]
I have initiated a discussion for decision on the Roslyn Compatibility Council mailing list.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -5011,4 +5011,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_TypeForwardedToMultipleAssemblies" xml:space="preserve">
     <value>Module '{0}' in assembly '{1}' is forwarding the type '{2}' to multiple assemblies: '{3}' and '{4}'.</value>
   </data>
+  <data name="ERR_PatternDynamicType" xml:space="preserve">
+    <value>It is not legal to use the type 'dynamic' in a pattern.</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: >in a pattern [](start = 53, length = 12)

Should we be more specific and say "in a declaration pattern" instead?

**Comment by AlekseyTs** [comment]
LGTM

**Comment by AlekseyTs** [comment]
Should this fix go to 15.1?

**Comment by gafter** [comment]
Whether or not this belongs in 15.1 depends on the guidelines for 15.1, which have not been shared with me yet. My understanding is that there will be a sweep of the PRs currently going into master to identify those that will be integrated into 15.1. In that case I'd rather they be integrated through that process (i.e. in order) to minimize conflicts.

**Pull request merged**